### ISSUE TITLE
QUESTION: EP and search ajax pagination inconsistencies

### ISSUE THREAD

**Issue by dashaluna** [10up/ElasticPress#2765]
Hello, I'm having issues (inconsistent results) while doing search page ajax pagination.

The current ElasticPress v.3.6.3 I have also tried this with the recent plugin v. 4.1.0  and the results are the same for both versions.

**Scenario:**
- I have several custom post types (that are searchable)
- I have WooCommerce and WooCommerce Membership plugins
- Posts per page is 10
- Search results page are using ajax pagination

### Page 1 (initial page, non-ajax):
When I search for ex. "surf" and hit search, the initial 1st page is loaded which is not an ajax request. I have 24 results.
The main WP_Query object is below:

```
WP_Query::__set_state( array(
	'query'                 => array(
		's' => 'surf',
	),
	'query_vars'            => array(
		's'                      => 'surf',
		'paged'                  => 0,
		...
		'ep_integrate'           => true,
		'ep_facet'               => true,
		'aggs'                   => array(
			'name'       => 'terms',
			'use-filter' => true,
			'aggs'       => array(
				...
			),
		),
		'posts_per_page'         => 10,
	...
	'request'               =>
			'SELECT SQL_CALC_FOUND_ROWS  wp_posts.* FROM wp_posts  WHERE 1=1  AND (((wp_posts.post_title LIKE \'{...}surf{51ba9ca71cb909d1004b412c4b7e11e2279c7e614317e99c433113373e3d5aa9}\')
			...
			DESC, wp_posts.post_date DESC LIMIT 0, 10',
	...
	'found_posts'           => 24,
	'max_num_pages'         => 3.0,
	...
	'is_paged'              => false,
	'is_admin'              => false,
	...
	'num_posts'             => 24,
	'elasticsearch_success' => true,
) );
```

### Page 2 - using ajax and `ep_ajax_wp_query_integration` filter
I then scroll down the search page, which kicks in the ajax request.
Reading ElasticPress docs, it does not work for ajax requests out of the box, so I had to add the `ep_ajax_wp_query_integration` filter, like so: `add_filter( 'ep_ajax_wp_query_integration', '__return_true' );`

In my ajax callback (in PHP) I construct a new query for the next page for search results, the WP_Query object is below:
```
WP_Query::__set_state( array(
	'query'                 => array(
		'posts_per_page' => 10,
		'paged'          => '2',
		'ep_integrate'   => true,
		's'              => 'surf',
	),
	'query_vars'            => array(
		'posts_per_page'         => 10,
		'paged'                  => 2,
		'ep_integrate'           => true,
		's'                      => 'surf',
		// NB: no `ep_facet`
		// NB: no `aggs`
		...
	'request'               =>
			'SELECT SQL_CALC_FOUND_ROWS  wp_posts.* FROM wp_posts  WHERE 1=1  AND (((wp_posts.post_title LIKE \'{67a237e8fe0f6060a58a4ffa0d322f65ee423cd5c513c87ec2d29c750ebd9029}surf{67a237e8fe0f6060a58a4ffa0d322f65ee423cd5c513c87ec2d29c750ebd9029}\')
			...
			// NB: plus extra WooCommerce Membership statuses
			... AND (wp_posts.post_status = \'publish\' OR ... wp_posts.post_status = \'wcm-active\' OR wp_posts.post_status = \'wcm-free_trial\' OR wp_posts.post_status = \'wcm-delayed\' OR wp_posts.post_status = \'wcm-complimentary\' OR wp_posts.post_status = \'wcm-pending\' OR wp_posts.post_status = \'wcm-paused\' OR wp_posts.post_status = \'wcm-expired\' OR wp_posts.post_status = \'wcm-cancelled\' OR wp_posts.post_author = ...
			DESC, wp_posts.post_date DESC LIMIT 10, 10',
	...
	'found_posts'           => 11,
	'max_num_pages'         => 2.0,
	...
	'is_paged'              => true,
	'is_admin'              => true,
	...
	'num_posts'             => 11,
	'elasticsearch_success' => true,
) );
```

However, I see some differences between the initial non-ajax WP_Query object:
- missing `ep_facet` key
- missing `aggs` key
- there are some extra WooCommerce Membership statuses in the `request`
- `is_admin => true`
- 'found_posts' => 11`

I'm guessing that Woo Membership statutes are added because there is `is_admin => true` which is always the case for ajax calls.

### Page 2 - using ajax and `ep_is_integrated_request` filter
I then decided to try the main `ep_is_integrated_request` filter instead of the `ep_ajax_wp_query_integration` one.
With this filter, I still get different results.

In my ajax callback (in PHP) the new WP_Query for the next page for search results is absolutely the same as in the step above. The only thing that's different is a different EP filter used this time.

```
WP_Query::__set_state( array(
	'query'                 => array(
		'posts_per_page' => 10,
		'paged'          => '2',
		'ep_integrate'   => true,
		's'              => 'surf',
	),
	'query_vars'            => array(
		'posts_per_page'         => 10,
		'paged'                  => 2,
		'ep_integrate'           => true,
		's'                      => 'surf',
		// NB: no `ep_facet`
		// NB: no `aggs`
		...
	'request'               =>
			'SELECT SQL_CALC_FOUND_ROWS  wp_posts.* FROM wp_posts  WHERE 1=1  AND (((wp_posts.post_title LIKE \'{5d7315c72822ea71ac73316ea375f05e3a9668c46db842e7b76c54a3e58a85f9}surf{5d7315c72822ea71ac73316ea375f05e3a9668c46db842e7b76c54a3e58a85f9}\')
			...
			// NB: plus extra WooCommerce Membership statuses
			... AND (wp_posts.post_status = \'publish\' OR ... wp_posts.post_status = \'wcm-active\' OR wp_posts.post_status = \'wcm-free_trial\' OR wp_posts.post_status = \'wcm-delayed\' OR wp_posts.post_status = \'wcm-complimentary\' OR wp_posts.post_status = \'wcm-pending\' OR wp_posts.post_status = \'wcm-paused\' OR wp_posts.post_status = \'wcm-expired\' OR wp_posts.post_status = \'wcm-cancelled\' OR wp_posts.post_author = ...
			DESC, wp_posts.post_date DESC LIMIT 10, 10',
	...
	'found_posts'           => 24,
	'max_num_pages'         => 3.0,
	...
	'is_paged'              => true,
	'is_admin'              => true,
	...
	'num_posts'             => 24,
	'elasticsearch_success' => true,
) );
```

**The main differences to the non-ajax inial WP_Query are:**
- the same differences as in the previous step for Page 2 ajax WP_Query with `ep_ajax_wp_query_integration` filter (as the new WP_Query did not change in this step).

**The main differences between Page 2 ajax WP_Query with `ep_ajax_wp_query_integration` filter:**
- interestingly, the number of the `found_posts` is different to when Page 2 ajax WP_Query is run with a different EP filter (in the step above).
- this time the `found_posts => 24` which I assume is correct as this is the same found posts number for the initial non-ajax query.
- why does the same ajax WP_Query return different `found_posts` results when different EP filters are used - I don't understand.

### Questions

1. What is the correct way to construct WP_Query for ajax pagination for search queries that run via EP?- I'd like to make sure that `request` does not include extra WooCommrce Membership statutes and all other query params are the same as for the non-ajax query.

2. Why am I seeing different `found_posts` results for the same WP_Query when I use different EP filters `ep_ajax_wp_query_integration` vs `ep_is_integrated_request`?

3. Is there a different way to do ajax pagination on the search page - still using EP/Elasticsearch?

Thank you.

**Comment by dashaluna** [comment]
PS I can see someone else also experienced the same issue with the search results ajax pagination as mentioned here https://github.com/10up/ElasticPress/issues/1082 but it wasn't resolved.

**Comment by oscarssanchez** [comment]
Hi @dashaluna ,

I'm curious to know the full query/query_vars. I see `'elasticsearch_success' => true` on both filter examples, so the only thing I can think at the moment is a different variable at play between these. I see you mention "WooCommerce Membership statuses", but I don't see them in the query, for example. Would you mind providing that as well? 

Page 1 query args seem correct if you have the facet feature enabled. If you are on a search/archive page and the feature is activated, these should be added to the query.

As far as an approach, what we usually do is create a new REST Endpoint instead of using Admin AJAX.

**Comment by dashaluna** [comment]
Hello @oscarssanchez thank you for getting back to me.

I'm not sure what else to provide for `WP_Query` objects, as you can see under the `query` and `query_vars` keys, there are the main query vars that are the same, like: `s`, `posts_per_page`, `ep_integrate`. Var `paged` is different as I'm requesting the next page of search results.

I've kept all the info that's different between `WP_Query` objects and removed everything that's the same. I've added WooCommerve Membership statutes into the `request`, I hope that helps. All that's been removed are the same between the objects.

---

As for the general approach for this, you mentioned that you create a new REST Endpoint instead of using WP ajax solution. 

Could you please point me to a tutorial to follow the steps on how to create and wire the parts for the REST Endpoint approach? Or specify the steps in this thread, please? Thank you!

**Comment by oscarssanchez** [comment]
Hi @dashaluna ,

Are those membership statuses then not being handled as a query arg? If that's the case, although it won't explain the difference in results, it is likely those are not being added.

I was trying to check a query with "surf" in our logs coming from your endpoint but I just was able to find 1 from 2 days ago, which seems odd since you probably have performed it more than that.

Could you please:

1.- Install the debug bar and debug bar elasticpress plugins
2.- Take the query arguments from all of your examples and for each create a new WP_Query
3.- Copy the query body/query results and paste them here.

If there's something else being added, I should be able to spot it by looking at the formatted JSON queries. 

As for the other approach, you still use AJAX, but the difference is you won't need to use wp-admin-ajax. Outlining briefly:

1.- Create a REST endpoint for pagination. You could make this endpoint just return search results or the whole html output as JSON.
2.- The callback function can take your search parameters and perform a normal WP_Query adding ep_integrate => true. This way you don't have to use any filters.

**Comment by oscarssanchez** [comment]
Hi @dashaluna ,

Just checking in, have you had a chance to look at my last reply?

**Comment by dashaluna** [comment]
Thank you @oscarssanchez  for looking into this and pointing me in the right direction.

---
I'm noting this here in case this helps someone else. I went ahead with a custom REST API endpoint solution.

On how to connect the different pieces (ajax call in a JS file + custom endpoint in a PHP file) these can help:
- https://www.excellarate.com/blogs/a-modern-way-of-ajax-in-wordpress-using-rest-api/ (recommended by @oscarssanchez )
- https://www.youtube.com/watch?v=rJH4cUhckhc (I found this one to follow on youtube. It's all done in one file, but that could be structured as required in different files, etc)
